src/fourierdomain/nuft/nuft.jl

```diff
@@ -8,21 +8,121 @@ Internal type used to store the cache for a non-uniform Fourier transform (NUFT)
 
 The user should instead create this using the [`FourierDualDomain`](@ref) function.
 """
-struct NUFTPlan{A,P,M} <: AbstractNUFTPlan
+struct NUFTPlan{A,P,M,I,T} <: AbstractNUFTPlan
     alg::A # which algorithm to use
     plan::P #NUFT matrix or plan
     phases::M #FT phases needed to phase center things
+    indices::I # imgdomain Ti/Fr indices mapped to visdomain indices
+    totalvis::T # Total number of visibility points
 end
 
+# This functions creates a tuple of vectors 1) what indices in imgdomain
+# correponds to 2) what all indices in the visdomain 
+# The indices tuple will be cached in NUFTPlan
+# Functions are overloaded based on the structure of RectiGrid
+function plan_indices(imgdomain::AbstractRectiGrid{<:Tuple{X,Y,Ti,Fr}},
+                      visdomain::UnstructuredDomain)
+    iminds = Tuple{Int,Int}[]
+    visinds = Vector{Int}[]
+    vis_points = domainpoints(visdomain)
+    Fr = imgdomain.Fr
+    Ti = imgdomain.Ti
+
+    for (j, fr) in pairs(Fr)
+        for (i, ti) in pairs(Ti)
+            push!(iminds, (i, j))
+            push!(visinds, findall(p -> p.Ti == ti && p.Fr == fr, vis_points))
+        end
+    end
+
+    return (iminds, visinds)
+end
+
+function plan_indices(imgdomain::AbstractRectiGrid{<:Tuple{X,Y,Fr,Ti}},
+                      visdomain::UnstructuredDomain)
+    iminds = Tuple{Int,Int}[]
+    visinds = Vector{Int}[]
+    vis_points = domainpoints(visdomain)
+    Ti = imgdomain.Ti
+    Fr = imgdomain.Fr
+
+    for (i, ti) in pairs(Ti)
+        for (j, fr) in pairs(Fr)
+            push!(iminds, (j, i))
+            push!(visinds, findall(p -> p.Ti == ti && p.Fr == fr, vis_points))
+        end
+    end
+    return (iminds, visinds)
+end
+
+function plan_indices(imgdomain::AbstractRectiGrid{<:Tuple{X,Y,Ti}},
+                      visdomain::UnstructuredDomain)
+    iminds = Int[]
+    visinds = Vector{Int}[]
+    vis_points = domainpoints(visdomain)
+    Ti = imgdomain.Ti
+
+    for (i, ti) in pairs(Ti)
+        push!(iminds, i)
+        push!(visinds, findall(p -> p.Ti == ti, vis_points))
+    end
+    return (iminds, visinds)
+end
+
+function plan_indices(imgdomain::AbstractRectiGrid{<:Tuple{X,Y,Fr}},
+                      visdomain::UnstructuredDomain)
+    iminds = Int[]
+    visinds = Vector{Int}[]
+    vis_points = domainpoints(visdomain)
+    Fr = imgdomain.Fr
+
+    for (j, fr) in pairs(Fr)
+        push!(iminds, j)
+        push!(visinds, findall(p -> p.Fr == fr, vis_points))
+    end
+    return (iminds, visinds)
+end
+
+# For only spatial case
+function plan_indices(imgdomain::AbstractRectiGrid{<:Tuple{X,Y}},
+                      visdomain::UnstructuredDomain)
+    return (0, 0)
+end
+
+# Function has been modified to process imgdomain/visdomain with Ti or Fr or both.
+# It calls the new plan_nuft  when Ti or Fr is present or else calls
+# the old spatial function
```

Comment:
We don't need these modified comments in the code. It will be pretty easy to forget when/why this was modified.